[controller] Fix LVMVolumeGroup spec.blockDeviceSelector update Issue

AleksZimin opened the pull request:
## Description
<!---
  Describe your changes in detail.

  Please let users know if your feature influences critical cluster components
  (restarts of ingress-controllers, control-plane, Prometheus, etc).
-->

This PR addresses an issue where the `spec.blockDeviceSelector` field of an existing `LVMVolumeGroup` resource was not updating during reconciliation, even when new physical volumes (PVs) were manually added on the node to the corresponding LVM Volume Group. With this fix, the `spec.blockDeviceSelector` now correctly reflects all associated `BlockDevice` resources, ensuring accurate representation of the volume group's state.

## Why do we need it, and what problem does it solve?
<!---
  This is the most important paragraph.
  You must describe the main goal of your feature.

  If it fixes an issue, place a link to the issue here.

  If it fixes an obvious bug, please tell users about the impact and effect of the problem.
-->

Previously, when new `BlockDevice` resources were manually introduced to an existing LVM Volume Group on a node, the `spec.blockDeviceSelector` field in the corresponding `LVMVolumeGroup` resource remained unchanged. This discrepancy could lead to mismanagement of storage resources and potential inconsistencies in volume group configurations. By ensuring that `spec.blockDeviceSelector` is updated appropriately, we maintain consistency between the actual state of the LVM Volume Group and its representation within Kubernetes.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

astef commented on images/agent/src/internal/controller/lvg/discoverer.go:
```diff
@@ -870,6 +882,51 @@ func configureBlockDeviceSelector(candidate internal.LVMVolumeGroupCandidate) *m
+				{
+					Key:      internal.MetadataNameLabelKey,
+					Operator: metav1.LabelSelectorOpIn,
+					Values:   blockDeviceNames,
```
Here, and on line 923, it's better to make a copy of the slice (`slices.Clone(blockDeviceNames)`), since in the scope of the function we don't know where did `blockDeviceName` came from and whether it's immutable.

asergunov replied:
Done

astef commented on images/agent/src/internal/controller/lvg/discoverer.go:
```diff
@@ -870,6 +882,51 @@ func configureBlockDeviceSelector(candidate internal.LVMVolumeGroupCandidate) *m
+
+	updated := false
+	found := false
+	for i := range existedLabelSelector.MatchExpressions {
```
Introduce loop variable and use it instead of repeating `existedLabelSelector.MatchExpressions[i]` each time

astef commented on images/agent/src/internal/controller/lvg/discoverer.go:
```diff
@@ -870,6 +882,51 @@ func configureBlockDeviceSelector(candidate internal.LVMVolumeGroupCandidate) *m
+	updated := false
+	found := false
+	for i := range existedLabelSelector.MatchExpressions {
+		if existedLabelSelector.MatchExpressions[i].Key == internal.MetadataNameLabelKey {
```
Inverse if statement to reduce nesting level

astef commented on images/agent/src/internal/controller/lvg/discoverer.go:
```diff
@@ -870,6 +882,51 @@ func configureBlockDeviceSelector(candidate internal.LVMVolumeGroupCandidate) *m
+				}
+			}
+		}
+	}
```
Here we assume that every expression has 'In' operator. What if there's a `NotIn` expression?

Also, there are cases, when device matches by labels different from `kubernetes.io/metadata.name` . In such cases we will add expressions unnecessarily.

I propose to use `k8s.io/apimachinery/pkg/labels` package to check if labels actually match.

asergunov commented:
This is a quick fix. We are going to implement proper one.

AleksZimin committed:
[controller] Fix LVMVolumeGroup spec.blockDeviceSelector update

Update spec.blockDeviceSelector during reconciliation when new
BlockDevices are added to the LVM Volume Group on the node, so the
field reflects the actual set of devices in the VG.

* Filter BlockDevices by node before computing candidates.
* Add notMatchedBlockDeviceNames helper that uses k8s.io/apimachinery
  labels selector to determine which device names are not matched.
* Add appendDeviceNamesToLabelSelector and updateBlockDeviceSelectorIfNeeded
  to extend an existing matchExpressions/matchLabels selector or create
  one from scratch when nil.
* Include selector mismatch in hasLVMVolumeGroupDiff so the
  discoverer triggers an update when the selector falls behind the
  actual device set.
* Add unit tests covering: no-op when up to date, creation from nil,
  appending to matchExpressions, migrating from matchLabels, and
  mixed selectors with unrelated keys.

Co-authored-by: Anton Sergunov <anton.sergunov@flant.com>
Co-authored-by: Aleksandr Stefurishin <aleksandr.stefurishin@flant.com>
Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>

duckhawk committed:
fix(lvg): actually persist Spec.BlockDeviceSelector update

The previous attempt called updateBlockDeviceSelectorIfNeeded(...) and
discarded the returned selector, then called only Status().Update(), so
the change to Spec.BlockDeviceSelector was never sent to the API server:

* When the existing selector was nil, the freshly-built selector was
  dropped on the floor (function return value assigned to "_").
* When the existing selector was non-nil, the in-memory mutation was
  performed but Status().Update() does not write the spec subresource,
  so the API state was never updated.

As a side-effect, hasLVMVolumeGroupDiff kept returning true on every
discoverer cycle (the selector mismatch never resolved), causing
UpdateLVMVolumeGroupByCandidate to run forever.

Fix:

* Capture the returned selector and assign it to lvg.Spec.BlockDeviceSelector
  when non-nil.
* Call cl.Update(ctx, lvg) to persist the spec change. controller-runtime
  refreshes lvg.ResourceVersion in place, so the subsequent
  Status().Update() picks up the new revision without a refetch.
* Move the spec Update before populating the status fields. The API
  server (and the fake client used in tests) preserves status on
  Update(), so doing the spec write after we'd populated lvg.Status
  would wipe those fields back to whatever the server had stored.

Add two regression tests:

* UpdateLVMVolumeGroup_persists_spec_blockDeviceSelector covers the
  "selector existed but is missing new devices" path: candidate brings
  three devices; we read the LVG back from the fake client and assert
  notMatchedBlockDeviceNames(stored.Spec.BlockDeviceSelector, ...)
  is empty, plus that status was also persisted.
* UpdateLVMVolumeGroup_creates_spec_blockDeviceSelector_when_nil
  covers the legacy LVG path: an LVG with nil selector gets a new one
  built from candidate.BlockDevicesNames.

Both regressions fail without the fix above.

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>
